### PR TITLE
[4401] Add router to Site Config and embedded mode

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/4401

### CODE REVIEW

File: templates/web/site-config.ftl [hunk 1]
```diff
@@ -58,7 +59,13 @@
 
 </head>
 <body class="yui-skin-cstudioTheme">
-<div id="admin-console" class="categories-panel-active"></div>
+  <#if RequestParameters.mode?? && RequestParameters.mode == "embedded">
+    <#assign mode = RequestParameters.mode />
+  <#else>
+    <#assign mode = ""/>
+  </#if>
```
Comment (rart) [inline templates/web/site-config.ftl]: You may want to put this above and, when mode is embedded, put some classes on the body and/or the root elem to target the sidebar & toolbar hide styles.

File: static-assets/themes/cstudioTheme/css/console.css [hunk 2]
```diff
@@ -116,6 +116,15 @@ input[type="checkbox"].content-type-selector--checkbox {
...
 }
 
+#admin-console.embedded #categories-panel{
+  display: none;
+}
+
+#admin-console.embedded #cstudio-admin-console-workarea {
+  width: 100%;
+  margin-left: 0;
+}
```
Comment (rart) [inline static-assets/themes/cstudioTheme/css/console.css]: Doesn't look like you're using these yet.

File: templates/web/site-config.ftl (hunk 1 above)
Comment (jvega190) [reply templates/web/site-config.ftl]: @rart right now the class if being set to the `admin-console` root element, and styles applied to that class

File: templates/web/site-config.ftl [hunk 3]
```diff
@@ -58,7 +59,13 @@
...
+
+  <div id="admin-console" class="categories-panel-active ${mode}"></div>
```
Comment (jvega190) [inline templates/web/site-config.ftl]: @rart here ^

File: static-assets/themes/cstudioTheme/css/console.css (hunk 2 above)
Comment (jvega190) [reply static-assets/themes/cstudioTheme/css/console.css]: The class is being set to the `#admin-console` element

File: templates/web/site-config.ftl (hunk 1 above)
Comment (jvega190) [reply templates/web/site-config.ftl]: I moved the class up to the body (I think it's clearer)

File: templates/web/site-config.ftl [hunk 4]
```diff
@@ -57,8 +58,14 @@
...
 </head>
-<body class="yui-skin-cstudioTheme">
-<div id="admin-console" class="categories-panel-active"></div>
+<#if RequestParameters.mode?? && RequestParameters.mode == "embedded">
+  <#assign mode = RequestParameters.mode />
+<#else>
+  <#assign mode = ""/>
+</#if>
+
+<body class="yui-skin-cstudioTheme ${mode}">
```
Comment (rart) [inline templates/web/site-config.ftl]: To make this safer, create a freemarker variable called "embedded" which you set to true if the query arg mode == embedded. Then, you set the embedded class name manually. This way there's no possibility of HTML injection.

File: templates/web/site-config.ftl [hunk 5]
```diff
@@ -42,6 +42,7 @@
 
   <script src="/studio/static-assets/libs/momentjs/moment.min.js?version=${UIBuildId!.now?string('Mddyyyy')}"></script>
   <script src="/studio/static-assets/libs/momentjs/moment-timezone-with-data-2012-2022.min.js?version=${UIBuildId!.now?string('Mddyyyy')}"></script>
+  <script src="/studio/static-assets/libs/routie/routie.min.js?version=${UIBuildId!.now?string('Mddyyyy')}"></script>
```
Comment (rart) [inline templates/web/site-config.ftl]: add routie version to its file name and remove the build num

File: ui/app/src/utils/i18n-legacy.ts [hunk 6]
```diff
@@ -405,6 +405,14 @@ const contentTypesMessages = defineMessages({
...
     id: 'siteConfig.controlNotAvailable',
     defaultMessage: 'Control not available'
+  },
+  loadModuleError: {
+    id: 'siteConfig.loadModuleError',
+    defaultMessage: 'Unable to load "{tool}". Check Site tools configuration.'
+  },
+  toolNotFound: {
+    id: 'siteConfig.toolNotFound',
+    defaultMessage: 'Tool not found.'
```
Comment (rart) [inline ui/app/src/utils/i18n-legacy.ts]: Include the tool id that wasn't found

File: static-assets/themes/cstudioTheme/css/console.css [hunk 7]
```diff
@@ -116,6 +116,15 @@ input[type="checkbox"].content-type-selector--checkbox {
...
 
+.embedded #admin-console #categories-panel{
+  display: none;
+}
+
+.embedded #admin-console #cstudio-admin-console-workarea {
+  width: 100%;
+  margin-left: 0;
+}
+
```
Comment (rart) [inline static-assets/themes/cstudioTheme/css/console.css:145]: There's a big margin at the top when the toolbar is removed. I don't see any styles addressing that; are you addressing somewhere else?

File: static-assets/components/cstudio-common/common-api.js [hunk 8]
```diff
@@ -6579,7 +6580,7 @@ var nodeOpen = false,
       /**
        * dynamically add a javascript file
        */
-      addJavascript: function(script, cache, callback) {
```
Comment (rart) [inline static-assets/components/cstudio-common/common-api.js:6582]: The change in the API is worrisome. Did you check the **entire** codebase for usages of addJavascript not using the original argument configuration? Did you searched for the "addJavascript" string alone or with its whole path?

Are cache and callback arts not used inside the function?

File: static-assets/components/cstudio-admin/base.js [hunk 9]
```diff
@@ -21,14 +21,20 @@
 
   CStudioAdminConsole = {
     toolContainerEls: [],
+    isEmbedded: false,
```
Comment (rart) [inline static-assets/components/cstudio-admin/base.js]: You could directly check here if the embedded class is present, right? No need to delay.

File: static-assets/components/cstudio-admin/base.js [hunk 10]
```diff
@@ -58,6 +64,38 @@
...
     },
 
+    initRouter(tools) {
+      const _self = this;
+      const toolsNames = tools.map((tool) => tool.name);
+
+      routie('tool/:name?', function(name) {
+        if (toolsNames.includes(name)) {
+          _self.currentRoute = name;
+          if (_self.toolsModules[name]) {
```
Comment (rart) [inline static-assets/components/cstudio-admin/base.js:72]: provided you already checked the name is correct above (line 72's if), is this if necessary?

File: static-assets/components/cstudio-admin/base.js [hunk 11]
```diff
@@ -58,6 +64,38 @@
...
+          elem.className = 'work-area-error';
+          $('#cstudio-admin-console-workarea').html(elem);
+          CrafterCMSNext.render(elem, 'ErrorState', {
+            graphicUrl: '/studio/static-assets/images/warning_state.svg',
+            classes: {
+              root: 'craftercms-error-state'
+            },
+            error: {
+              code: '',
+              message: formatMessage(contentTypesMessages.toolNotFound)
```
Comment (rart) [inline static-assets/components/cstudio-admin/base.js]: Include tool id in error message

File: static-assets/components/cstudio-admin/base.js [hunk 12]
```diff
@@ -95,11 +143,24 @@
...
               toolContainerEl: toolContainerEl
             };
-
+            const tool = config.tools.tool[j];
             CStudioAuthoring.Module.requireModule(
-              'cstudio-console-tools-' + config.tools.tool[j].name,
-              '/static-assets/components/cstudio-admin/mods/' + config.tools.tool[j].name + '.js',
-              { config: config.tools.tool[j] },
+              'cstudio-console-tools-' + tool.name,
+              '/static-assets/components/cstudio-admin/mods/' + tool.name + '.js',
```
Comment (rart) [inline static-assets/components/cstudio-admin/base.js]: Use modern syntax for string interpolation

**Review by rart — CHANGES_REQUESTED** [review]

File: static-assets/components/cstudio-admin/base.js (hunk 10 above)
Comment (jvega190) [reply static-assets/components/cstudio-admin/base.js:72]: @rart yes it is necessary, the toolsNames list has the names (ids) of the tools that are to be loaded, the toolsModules are the ones already loaded

File: static-assets/components/cstudio-common/common-api.js (hunk 8 above)
Comment (jvega190) [reply static-assets/components/cstudio-common/common-api.js:6582]: I searched for the string `addJavascript` (alone) all over the code, and I found no usages with the optional arguments, also, in the code neither cache nor callback are being used

**Review by rart — APPROVED** [review]

**Pull request merged**